pkg/reconciler/trait/reconciler.go

```diff
@@ -0,0 +1,227 @@
+	errGetPackage             = "cannot get package for workload reference in trait"
+	errApplyTraitModification = "cannot apply trait modification to workload package"
+)
+
+// Reconcile event reasons.
+const (
+	reasonTraitWait   = "WaitingForWorkloadPackage"
+	reasonTraitModify = "PackageModified"
+
+	reasonCannotGetPackage        = "CannotGetReferencedWorkloadPackage"
+	reasonCannotModifyPackage     = "CannotModifyPackage"
+	reasonCannotApplyModification = "CannotApplyModification"
+)
+
+// A ReconcilerOption configures a Reconciler.
+type ReconcilerOption func(*Reconciler)
+
+// WithLogger specifies how the Reconciler should log messages.
+func WithLogger(l logging.Logger) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.log = l
+	}
+}
+
+// WithRecorder specifies how the Reconciler should record events.
+func WithRecorder(er event.Recorder) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.record = er
+	}
+}
+
+// WithModifier specifies how the Reconciler should modify the workload package.
+func WithModifier(m Modifier) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.trait = m
+	}
+}
+
+// WithAccessors specifies how the Reconciler should access the part of the
+// workload package to be modified.
+func WithAccessors(m ModifyAccessor) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.accessor = m
+	}
+}
+
+// WithApplicator specifies how the Reconciler should apply the workload
+// package modification.
+func WithApplicator(a resource.Applicator) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.applicator = a
+	}
+}
+
+// A Reconciler reconciles OAM traits by modifying the the KubernetesApplication
+// that packages their referenced workload.
+type Reconciler struct {
+	client     client.Client
+	newTrait   func() Trait
+	newPackage func() Object
+	trait      Modifier
+	accessor   ModifyAccessor
+	applicator resource.Applicator
+
+	log    logging.Logger
+	record event.Recorder
+}
+
+// Kind is an OAM trait kind.
+type Kind schema.GroupVersionKind
+
+// A WorkloadReferencer is a type that has a Workload reference.
+type WorkloadReferencer interface {
+	GetWorkloadReference() oamv1alpha2.WorkloadReference
+	SetWorkloadReference(oamv1alpha2.WorkloadReference)
+}
+
+// A Trait is a type of OAM trait.
+type Trait interface {
+	resource.Conditioned
+	metav1.Object
+	runtime.Object
+	WorkloadReferencer
+}
+
+// Object is a KubernetesObject
+type Object interface {
+	metav1.Object
+	runtime.Object
+}
+
+// NewReconciler returns a Reconciler that reconciles OAM traits by fetching
+// their referenced workload's KubernetesApplication and applying modifications.
+func NewReconciler(m ctrl.Manager, trait Kind, p Kind, o ...ReconcilerOption) *Reconciler {
+	nt := func() Trait {
+		return resource.MustCreateObject(schema.GroupVersionKind(trait), m.GetScheme()).(Trait)
+	}
+
+	np := func() Object {
+		return resource.MustCreateObject(schema.GroupVersionKind(p), m.GetScheme()).(Object)
+	}
```

Comment:
I might be misinterpreting suggestion here, but are they not currently being called inside `NewReconciler`?

Reply:
```go
_, _ = nt(), np()
```

We use the above convention in other similar controllers immediately after declaring the `nt` and `np` functions, inside `NewReconciler`. Currently `NewReconciler` never calls these functions, but instead passes them to the `*Reconciler` to be called at reconcile time.

`resource.MustCreateObject` will panic if the supplied GVK is not registered with the supplied scheme. If that's going to happen we'd ideally notice it sooner, when `NewReconciler` was called (typically during controller manager startup) rather than later, the first time we actually try to reconcile a resource.

Reply:
I'd also recommend including a comment briefly communicating why we call these functions. Folks familiar with Go idioms may spot that `MustFoo` typically means "this will panic if it fails", but to others it's not very intuitive.

Reply:
Ah yes, my apologies I did not notice that I had missed that! thank you :)

Reply:
Fixed!